dalle2_pytorch/dataloaders/decoder_loader.py

```diff
@@ -0,0 +1,208 @@
+import os
+import webdataset as wds
+import torch
+import numpy as np
+import fsspec
+
+def get_shard(filename):
+    """
+    Filenames with shards in them have a consistent structure that we can take advantage of
+    Standard structure: path/to/file/prefix_string_00001.ext
+    """
+    try:
+        return int(filename.split("_")[-1].split(".")[0])
+    except ValueError:
+        raise RuntimeError(f"Could not find shard for filename {filename}")
+
+# fsspec file reading code from embedding_reader https://github.com/rom1504/embedding-reader/blob/dd7da26ba4cd363cbfd608e0877faa921180c2e3/embedding_reader/get_file_list.py
+def get_file_list(path, file_format):
+    """
+    Get the file system and all the file paths that matches `file_format` under the given `path`.
+    The `path` could a single folder or multiple folders.
+    :raises ValueError: if file system is inconsistent under different folders.
+    """
+    if isinstance(path, str):
+        return _get_file_list(path, file_format)
+    all_file_paths = []
+    fs = None
+    for p in path:
+        cur_fs, file_paths = _get_file_list(p, file_format, sort_result=False)
+        if fs is None:
+            fs = cur_fs
+        elif fs != cur_fs:
+            raise ValueError(
+                f"The file system in different folder are inconsistent.\n" f"Got one {fs} and the other {cur_fs}"
+            )
+        all_file_paths.extend(file_paths)
+    all_file_paths.sort()
+    return fs, all_file_paths
+
+def make_path_absolute(path: str) -> str:
+    fs, p = fsspec.core.url_to_fs(path)
+    if fs.protocol == "file":
+        return os.path.abspath(p)
+    return path
+
+def _get_file_list(
+    path: str, file_format: str, sort_result: bool = True
+):
+    """Get the file system and all the file paths that matches `file_format` given a single path."""
+    path = make_path_absolute(path)
+    fs, path_in_fs = fsspec.core.url_to_fs(path)
+    prefix = path[: path.index(path_in_fs)]
+    glob_pattern = path.rstrip("/") + f"/**.{file_format}"
+    file_paths = fs.glob(glob_pattern)
+    if sort_result:
+        file_paths.sort()
+    file_paths_with_prefix = [prefix + file_path for file_path in file_paths]
+    return fs, file_paths_with_prefix
+### Code from embedding_reader ends
+
+def embedding_inserter(samples, embeddings_url, shard_width, handler=wds.handlers.reraise_exception):
+    """Given a datum of {"__key__": str, "__url__": str, ...} adds the cooresponding embedding and yields"""
+    previous_tar_url = None
+    current_embeddings = None
+    # Get a reference to an abstract file system where the embeddings are stored
+    fs, embedding_files = get_file_list(embeddings_url, 'npy')
```

Comment:
Do we need the full list here ? It seems you're only using the first item

Reply:
Ah, that's relic of using a different method of finding the corresponding embedding file. If I change this to a simpler method of getting the paths, it would also mean I could remove the heavy functions above and simplify the thing a bit at the cost of a little bit of safety.

Reply:
Check the newest commit for this change.